File: swarmmanagercontroller/src/main/java/co/uk/swarmbit/docker/api/nodes/NodesApiImpl.java
package co.uk.swarmbit.docker.api.nodes;

import co.uk.swarmbit.docker.api.common.AbstractApiImpl;
import co.uk.swarmbit.docker.api.nodes.parameters.NodeDeleteParameters;
import co.uk.swarmbit.docker.api.nodes.parameters.NodeUpdateParameters;
import co.uk.swarmbit.docker.api.nodes.parameters.NodesListParameters;
import co.uk.swarmbit.docker.api.common.json.NodeJson;
import co.uk.swarmbit.rest.RestResponseType;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class NodesApiImpl extends AbstractApiImpl implements NodesApi {

    private static final String NODES_PATH = "/nodes";

    private static final String UPDATE_PATH = "/update";

    @Override
    public List<NodeJson> listNodes(String swarmId, NodesListParameters parameters) {
        return listObjects(NODES_PATH, swarmId, new RestResponseType<List<NodeJson>>() {}, parameters);
    }

    @Override
    public NodeJson inspectNode(String swarmId, String id) {
        return  inspectObject(NODES_PATH, swarmId, new RestResponseType<NodeJson>() {}, id);
    }

    @Override
    public void updateNode(String swarmId, String id, NodeUpdateParameters parameters) {
        updateObject(NODES_PATH + "/" + id + UPDATE_PATH, swarmId, new RestResponseType<Void>() {}, parameters, parameters);
    }

    @Override
    public void deleteNode(String swarmId, String id, NodeDeleteParameters parameters) {
        deleteObject(NODES_PATH + "/" + id, swarmId, new RestResponseType<Void>(){}, parameters);
    }

}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/nodes/NodesApiImpl.kt
package com.swarmbit.docker.api.nodes

import com.swarmbit.docker.api.common.AbstractApiImpl
import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.rest.model.ResponseType
import com.swarmbit.docker.api.nodes.model.Node
import com.swarmbit.docker.api.nodes.parameters.NodesDeleteParameters
import com.swarmbit.docker.api.nodes.parameters.NodesListParameters
import com.swarmbit.docker.api.nodes.parameters.NodesUpdateParameters
import io.micronaut.context.annotation.Context

@Context
class NodesApiImpl(dockerWebClient: DockerWebClient) : AbstractApiImpl(dockerWebClient), NodesApi {
    companion object {
        private const val NODES_PATH = "/nodes"
    }

    override fun listNodes(swarmId: String, parameters: NodesListParameters): List<Node> {
        return listObjects(NODES_PATH, swarmId, object : ResponseType<List<Node>>() {}, parameters)
    }

    override fun inspectNode(swarmId: String, id: String): Node {
        return inspectObject(NODES_PATH, swarmId, object : ResponseType<Node>() {}, id)
                ?: throw IllegalArgumentException("No node found for swarmId ($swarmId) and nodeId (${id})")
    }

    override fun updateNode(swarmId: String, id: String, parameters: NodesUpdateParameters) {
        updateObject("$NODES_PATH/$id", swarmId, object : ResponseType<Void>() {}, parameters)
                ?: throw IllegalArgumentException("No node found for swarmId ($swarmId) and nodeId (${id})")
    }

    override fun deleteNode(swarmId: String, id: String, parameters: NodesDeleteParameters) {
        deleteObject("$NODES_PATH/$id", swarmId, object : ResponseType<Void>() {}, parameters)
    }

}